### PR TITLE
feat(RingTheory.DedekindDomain.FiniteAdeleRing): add topology

### PR DESCRIPTION
**Pull request by mariainesdff**
We endow the finite adele ring with the restricted product topology and show that it makes it into a topological ring.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by Rida-Hamadani — COMMENTED** [review]
Style

**Comment by kbuzzard** [comment]
Yikes, the proofs that addition and multiplication are continuous are *long*! 

I wonder if we can somehow use results in https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Nonarchimedean/Bases.html to golf these? I'm actively looking into this.

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean [hunk 1]
```diff
@@ -336,6 +335,306 @@ instance : Coe (FiniteAdeleRing R K) (K_hat R K) where
...
+      (fun p : v.adicCompletion K × v.adicCompletion K => p.fst + p.snd) ⁻¹' V v)
+    {Vx : ∀ v : HeightOneSpectrum R, Set (v.adicCompletion K)}
+    (hVx : Vx = fun v => ite (x.val v ∈ v.adicCompletionIntegers K ∧
+        y.val v ∈ v.adicCompletionIntegers K ∧ V v = v.adicCompletionIntegers K)
+      (v.adicCompletionIntegers K : Set (HeightOneSpectrum.adicCompletion K v))
+      (choose (isOpen_prod_iff.mp (hV v) (x.val v) (y.val v) (hxy v)))) :
+    IsOpen {z : FiniteAdeleRing R K | ∀ v : HeightOneSpectrum R, z.val v ∈ Vx v} := by
+  apply TopologicalSpace.isOpen_generateFrom_of_mem
+  use Vx
+  refine' ⟨fun x => by rfl, _, _⟩
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:416]: I could add: all `refine'` in mathlib has been removed. As I understand it, there's no consensus yet as to whether to disallow it entirely - but if `refine` just does it, that is quite clearly preferred.

**Comment by kbuzzard** [comment]
Closed in favour of #13705 (WIP for now)

**Pull request closed without merging**